File: Go/add-one-row-to-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    if depth == 1 {
        return &TreeNode{val, root, nil}
    }
    if depth == -1 {
        return &TreeNode{val, nil, root}
    }
    if root != nil {
        root.Left = addOneRow(root.Left, val, abs(depth) - 1)
        root.Right = addOneRow(root.Right, val, -(abs(depth) - 1))
    }
    return root
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

File: TypeScript/add-one-row-to-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function addOneRow(root: TreeNode | null, val: number, depth: number): TreeNode | null {
    return helper(root, val, depth, null);
}

function helper(root: TreeNode | null, val: number, depth: number, isLeft: boolean | null): TreeNode | null {
    if (depth === 1) {
        if (isLeft === null || isLeft === true) {
            return new TreeNode(val, root, null);
        } else {
            return new TreeNode(val, null, root);
        }
    }
    if (root) {
        root.left = helper(root.left, val, depth - 1, true);
        root.right = helper(root.right, val, depth - 1, false);
    }
    return root;
};